src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs

```diff
@@ -1443,231 +1465,202 @@ public StringBuilder AppendFormat(IFormatProvider? provider, [StringSyntax(Strin
-                            break;
+                            ThrowHelper.ThrowFormatInvalidString();
                         }
-                        else if (ch == '{')
+                        ch = MoveNext(format, ref pos);
+                        while (char.IsAsciiDigit(ch) && width < WidthLimit)
                         {
-                            // Braces inside the argument hole are not supported
-                            FormatError();
+                            width = width * 10 + ch - '0';
+                            ch = MoveNext(format, ref pos);
                         }
 
-                        pos++;
+                        // Consume optional whitespace
+                        while (ch == ' ')
+                        {
+                            ch = MoveNext(format, ref pos);
+                        }
                     }
 
-                    if (pos > startPos)
+                    // The next character needs to either be a closing brace for the end of the hole,
+                    // or a colon indicating the start of the format.
+                    if (ch != '}')
                     {
-                        itemFormatSpan = format.AsSpan(startPos, pos - startPos);
+                        if (ch != ':')
+                        {
+                            // Unexpected character
+                            ThrowHelper.ThrowFormatInvalidString();
+                        }
+
+                        // Search for the closing brace; everything in between is the format,
+                        // but opening braces aren't allowed.
+                        int startingPos = pos;
+                        while (true)
+                        {
+                            ch = MoveNext(format, ref pos);
+
+                            if (ch == '}')
+                            {
+                                // Argument hole closed
+                                break;
+                            }
+
+                            if (ch == '{')
+                            {
+                                // Braces inside the argument hole are not supported
+                                ThrowHelper.ThrowFormatInvalidString();
+                            }
+                        }
+
+                        startingPos++;
+                        itemFormatSpan = format.AsSpan(startingPos, pos - startingPos);
                     }
                 }
-                else if (ch != '}')
-                {
-                    // Unexpected character
-                    FormatError();
-                }
 
                 // Construct the output for this arg hole.
+                Debug.Assert(format[pos] == '}');
                 pos++;
                 string? s = null;
                 string? itemFormat = null;
 
+                if ((uint)index >= (uint)args.Length)
+                {
+                    throw new FormatException(SR.Format_IndexOutOfRange);
```

Comment:
Why no throw-helper here?

Reply:
That makes sense.
I just wondered about consistency with the other places (not arguing that because of this a throw-helper should be used here).